### PR TITLE
ENH: Add the decimal.Decimal type to infer_dtypes (#15690)

### LINKED ISSUE #15690: Infer Decimal dtype
```python
In [15]: from decimal import Decimal

In [16]: s = pd.Series([Decimal(3), Decimal(1), Decimal(4)])
    ...: s
Out[16]: 
0    3
1    1
2    4
dtype: object

In [17]: s.dtype
Out[17]: dtype('O')

In [18]: pd.lib.infer_dtype(s)
Out[18]: 'mixed'

In [19]: set(map(type, s))
Out[19]: {decimal.Decimal}
```
It would be nice if the `infer_dtype` function could determine if the column was all of decimal type which could be useful when working with databases which return data as decimal types

### PR DESCRIPTION
**Pull request by margaret**
- [x] closes #15690 
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/dtypes/test_inference.py [hunk 1]
```diff
@@ -462,6 +463,11 @@ def test_floats(self):
         result = lib.infer_dtype(arr)
         assert result == 'floating'
 
+    def test_decimals(self):
+        arr = np.array([Decimal(1), Decimal(2), Decimal(3)])
```
Comment (TomAugspurger) [inline pandas/tests/dtypes/test_inference.py]: Can you add a comment with a link to the github issue? (#15690)

File: pandas/tests/dtypes/test_inference.py [hunk 2]
```diff
@@ -462,6 +463,12 @@ def test_floats(self):
         result = lib.infer_dtype(arr)
         assert result == 'floating'
 
+    def test_decimals(self):
+        # GH15690
+        arr = np.array([Decimal(1), Decimal(2), Decimal(3)])
+        result = lib.infer_dtype(arr)
+        assert result == 'decimal'
+
```
Comment (TomAugspurger) [inline pandas/tests/dtypes/test_inference.py]: Sorry, could you add a test with a mix of Decimal and non-decimal (like floats) and make sure that the return is `mixed`? Or search through and see if we have a test that already covers this. Other than that, this looks great.

**Comment by jorisvandenbossche** [comment]
What I am wondering is what are potential consequences of this change. As there are quite some places in the codebase where we wheck if the inferred type is 'mixed', which could lead to potential changes in behaviour for objects with Decimals with this change. 
But don't know if those cases in the code would be relevant for Decimal objects.

**Comment by margaret** [comment]
https://travis-ci.org/pandas-dev/pandas/jobs/234948322#L1392 failed one of the vector resizing tests (`Test for memory errors after internal vector reallocations`), but only on the Python 3.6 run. Any ideas on how that would be related to this? The parameter it failed on seems to just be int64.

**Comment by TomAugspurger** [comment]
@jorisvandenbossche are there places where we explicitly check `== 'mixed'`? I haven't been able to find in just grepping (other than tests).

@margaret I'm not able to reproduce that. I'll restart that worker.

**Comment by jorisvandenbossche** [comment]
According to my pycharm search, there are 13 occurrences of 'mixed' (usage as a string constant)

File: doc/source/whatsnew/v0.21.0.txt [hunk 3]
```diff
@@ -34,6 +34,7 @@ Other Enhancements
 - ``Series.to_dict()`` and ``DataFrame.to_dict()`` now support an ``into`` keyword which allows you to specify the ``collections.Mapping`` subclass that you would like returned.  The default is ``dict``, which is backwards compatible. (:issue:`16122`)
 - ``RangeIndex.append`` now returns a ``RangeIndex`` object when possible (:issue:`16212`)
 - :func:`to_pickle` has gained a protocol parameter (:issue:`16252`). By default, this parameter is set to `HIGHEST_PROTOCOL <https://docs.python.org/3/library/pickle.html#data-stream-format>`__
+- ``lib.infer_dtype`` now infers decimals. (:issue: `15690`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: ``:func:`api.types.infer_dtype` ``

File: pandas/_libs/src/inference.pyx [hunk 4]
```diff
@@ -308,7 +308,6 @@ def infer_dtype(object value):
     'categorical'
 
     """
```
Comment (jreback) [inline pandas/_libs/src/inference.pyx]: pls update the doc-string & add an example

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
If the suite passes this is prob ok. The times when we infer are: 

1) construction (IOW we have something like ``Series([Decimal(..), Decimal(..)])``, but since ``decimal`` is an object dtype this won't matter (IOW, we don't specially handle it so it will fall thru as object anyhow).

2) indexing, with an *Index* of decimals. but again its the same as above, we don't do anything special. so I think ok for now.

**Comment by jorisvandenbossche** [comment]
One further thought: would we like to call this `"mixed-decimal"` ? (or "object-decimal", but we seem to use 'mixed' as general indicator for object dtyped columns) 
This is more explicit in the fact that it is still object dtype (since we don't have a decimal dtype, nor do we plan to have one), but lets you specifically check for decimals, and also let's you more easily test for object dtypes in general (`"mixed" in inferred_dtype`, instead of `inferred_dtype` in ['decimal', 'mixed', 'mixed-..', ..]`)

File: pandas/_libs/src/inference.pyx (hunk 4 above)
Comment (margaret) [inline pandas/_libs/src/inference.pyx]: https://github.com/pandas-dev/pandas/pull/16426/commits/6f158cb23cbfc51b35f9109a65e2929721f1e9a2

**Comment by jreback** [comment]
i think decimal is just fine ; don't need to get more complicated

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by jreback** [comment]
thanks!

**Pull request merged**